### ISSUE TITLE
AUTH_LDAP_APPEND_DOMAIN doesnt' work with Active Directory

### ISSUE THREAD

**Issue by AlexisLessard** [dpgaspar/Flask-AppBuilder#1123]
### Environment

Flask-Appbuilder version: 1.13.1

pip freeze output:
alembic==1.1.0
amqp==2.5.1
apache-airflow==1.10.5
apispec==2.0.2
attrs==19.1.0
Babel==2.7.0
billiard==3.6.1.0
cached-property==1.5.1
celery==4.3.0
certifi==2019.9.11
chardet==3.0.4
Click==7.0
colorama==0.4.1
colorlog==4.0.2
configparser==3.5.3
croniter==0.3.30
defusedxml==0.6.0
dill==0.2.9
docutils==0.15.2
dumb-init==1.2.2
Flask==1.1.1
Flask-Admin==1.5.3
Flask-AppBuilder==1.13.1
Flask-Babel==0.12.2
Flask-Caching==1.3.3
Flask-JWT-Extended==3.23.0
Flask-Login==0.4.1
Flask-OpenID==1.2.5
Flask-SQLAlchemy==2.4.0
flask-swagger==0.2.13
Flask-WTF==0.14.2
flower==0.9.3
funcsigs==1.0.0
future==0.16.0
gunicorn==19.9.0
idna==2.8
iso8601==0.1.12
itsdangerous==1.1.0
Jinja2==2.10.1
json-merge-patch==0.2
jsonschema==3.0.2
kombu==4.6.3
lazy-object-proxy==1.4.2
ldap3==2.6.1
lockfile==0.12.2
Mako==1.1.0
Markdown==2.6.11
MarkupSafe==1.1.1
marshmallow==2.19.5
marshmallow-enum==1.5.1
marshmallow-sqlalchemy==0.19.0
numpy==1.17.2
ordereddict==1.1
pandas==0.25.1
pendulum==1.4.4
prison==0.1.0
psutil==5.6.3
psycopg2-binary==2.8.3
pyasn1==0.4.7
pyasn1-modules==0.2.6
Pygments==2.4.2
PyJWT==1.7.1
pyrsistent==0.15.4
python-daemon==2.1.2
python-dateutil==2.8.0
python-editor==1.0.4
python-ldap==3.2.0
python3-openid==3.1.0
pytz==2019.2
pytzdata==2019.2
PyYAML==5.1.2
redis==3.3.8
requests==2.22.0
setproctitle==1.1.10
six==1.12.0
SQLAlchemy==1.3.8
tabulate==0.8.3
tenacity==4.12.0
termcolor==1.1.0
text-unidecode==1.2
thrift==0.11.0
tornado==5.1.1
tzlocal==1.5.1
unicodecsv==0.14.1
urllib3==1.25.3
vine==1.3.0
Werkzeug==0.15.6
WTForms==2.2.1
zope.deprecation==4.4.0

Python: 3.7.4

### Describe the expected results

When configuring AUTH_LDAP_APPEND_DOMAIN (let's say,` test.com`) in my webserver_config.py file and when connection against an Active Directory server, I want to be able to connect without having to add `test.com` after my username. 

### Describe the actual results

However it never works, because the username returned by [`con.search_s`here](https://github.com/dpgaspar/Flask-AppBuilder/blob/v1.13.1/flask_appbuilder/security/manager.py#L719) is a DN, and not juste the username. This results in [`con.bind_s(username, password)` here](https://github.com/dpgaspar/Flask-AppBuilder/blob/v1.13.1/flask_appbuilder/security/manager.py#L770) trying to bind itself with the username `CN=Alexis Lessard,OU=Somewhere,DC=test,DC=com@test.com`, which is wrong.
Without the option, airflow has no problems to authenticate when inputing the domain in the login field, because it works as a whole with values it understands (the result of an AD search is then tested against an AD server).  

### Steps to reproduce

1. Have an Microsoft Active Directory server available, with groups and users available with an ldaps url.
2. Install apache-airflow 1.10.5 and enable rbac in your airflow.cfg file
3. Configure your ldap authentification in your webservr_config.py file like this (values are changed):
```
AUTH_TYPE = AUTH_LDAP
AUTH_LDAP_SERVER = "ldaps://blabla.test:636"
AUTH_LDAP_TLS_DEMAND = True
AUTH_LDAP_BIND_USER = "CN=bob,DC=test"
AUTH_LDAP_BIND_PASSWORD = "tis_a_secret"
AUTH_LDAP_TLS_CACERTFILE = "/home/airflow/ca.crt"
AUTH_LDAP_UID_FIELD = "userPrincipalName"
AUTH_LDAP_SEARCH = "OU=Where_your_users_are_stored,DC=test, DC=com"
AUTH_LDAP_SEARCH_FILTER = "memberOf:1.2.840.113556.1.4.1941:=CN=Group,OU=Where_it_is,DC=test,DC=com"
AUTH_LDAP_APPEND_DOMAIN = test.com
```

Possibilty, there may be a better uid field, one that only includes the login name. I'll try to work in that direction. In the meantime, I propose a setting like AUTH_LDAP_ACTIVE_DIRECTORY, that would change some settings to their correct value, and treat the returned username correctly.

**Comment by AlexisLessard** [comment]
Tried replacing `AUTH_LDAP_UID_FIELD` by `sAMAccountName`, as this field was used in the older way of authenticating to apache airflow. No user is found, wether I use `AUTH_LDAP_APPEND_DOMAIN` or not.

FYI, I'm adding a bunch of log messages ad hoc in my library to see what values are used by the application, and where the code breaks. There are some debug log messages by default, but they don't seem to appear in my logs. Is there a `webserver_config.py` setting to change the log level of AppBuilder?

**Comment by AlexisLessard** [comment]
The user returned by `_search_ldap` is of this format:
```
[
    (
        'CN=Alexis Lessard,OU=Where_your_users_are_stored,DC=test,DC=com', 
        {
            'sn': [b'Lessard'],
            'givenName': [b'Alexis'],
            'mail': [b'alexis.lessard@test.com']
        }
    )
]
```
Therefore, the username resulting in `user[0][0]` [here](https://github.com/dpgaspar/Flask-AppBuilder/blob/v1.13.1/flask_appbuilder/security/manager.py#L762) results in this:
```
CN=Alexis Lessard,OU=Where_your_users_are_stored,DC=test,DC=com
```
to which we append `@test.com` with the `AUTH_LDAP_APPEND_DOMAIN` setting set. Can't authenticate with that, and there's nothing to filter out, so I don't think `AUTH_LDAP_USERNAME_FORMAT` can be useful here. But if there was a way to skip appending the domain for the binding process when we are authentication against an AD server... This is where an `AUTH_LDAP_ACTIVE_DIRECTORY` setting could be useful. I'll do a test ad hoc.

**Comment by dpgaspar** [comment]
On ActiveDirectory I've used the following config:
```
AUTH_LDAP_UID = 'userPrincipalName'
AUTH_LDAP_APPEND_DOMAIN = 'miniclip.com'
```

**Comment by AlexisLessard** [comment]
@dpgaspar, whatversion of AppBuilder are you using?I'm using 1.13.1 since it's the one that  comes with apache airflow, but maybe the issue was resolved later. Maybe your AD configuration is different than mine, too. To be sure, I'd like to see what's the content of the `username` variable when you get [here](https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/manager.py#L825)

Also, where is `AUTH_LDAP_UID` defined? In flask_appbuilder/security/manager.py, the setting I see is `AUTH_LDAP_UID_FIELD`.

**Comment by dpgaspar** [comment]
I used this on a 1.X.X version. the `givenName` should work also (note that it's case sensitive).

btw, by commenting:
```
#    if self.auth_ldap_append_domain:
#        username = username + '@' + self.auth_ldap_append_domain
```
has the same effect of `AUTH_LDAP_APPEND_DOMAIN=""` or simply not declaring it, right?

**Comment by AlexisLessard** [comment]
@dpgaspar, no, because the setting is used [here](https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/manager.py#L766-L767) to search for the username on AD. That way. the correct user is returned without specifying the domain at login, but the username returnd is in a DN format (view earlier), which won't work when appending the domain at the bind step [here](https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/manager.py#L827). 

On a side note, your code is really easy to understand and reverse engineer! That's how I was able to dig so deep into this issue. Nice!

**Comment by dpgaspar** [comment]
Thank you, but I personally really don't like this LDAP auth part, really need to refactor it. Stupid question, what happens if you remove `AUTH_LDAP_SEARCH` and `AUTH_LDAP_SEARCH_FILTER`

**Comment by AlexisLessard** [comment]
```pytb
Traceback (most recent call last):
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask_appbuilder/security/manager.py", line 826, in auth_user_ldap
    if not self._bind_ldap(ldap, con, username, password):
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask_appbuilder/security/manager.py", line 758, in _bind_ldap
    user = self._search_ldap(ldap, con, username)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask_appbuilder/security/manager.py", line 726, in _search_ldap
    self.auth_ldap_email_field
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 854, in search_s
    return self.search_ext_s(base,scope,filterstr,attrlist,attrsonly,None,None,timeout=self.timeout)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 848, in search_ext_s
    return self.result(msgid,all=1,timeout=timeout)[1]
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 740, in result
    resp_type, resp_data, resp_msgid = self.result2(msgid,all,timeout)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 744, in result2
    resp_type, resp_data, resp_msgid, resp_ctrls = self.result3(msgid,all,timeout)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 751, in result3
    resp_ctrl_classes=resp_ctrl_classes
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 758, in result4
    ldap_result = self._ldap_call(self._l.result4,msgid,all,timeout,add_ctrls,add_intermediates,add_extop)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 331, in _ldap_call
    reraise(exc_type, exc_value, exc_traceback)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/compat.py", line 44, in reraise
    raise exc_value
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/ldap/ldapobject.py", line 315, in _ldap_call
    result = func(*args,**kwargs)
ldap.NO_SUCH_OBJECT: {'desc': 'No such object', 'info': "0000208D: NameErr: DSID-031001EE, problem 2001 (NO_OBJECT), data 0, best match of:\n\t''\n"}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask_appbuilder/security/views.py", line 496, in login
    form.password.data
  File "/opt/pgw/virtualenv/lib/python3.7/site-packages/flask_appbuilder/security/manager.py", line 867, in auth_user_ldap
    if type(e.message) == dict and 'desc' in e.message:
AttributeError: 'NO_SUCH_OBJECT' object has no attribute 'message'
```

The issue fixed in #1118 occurs :P But I can't apply the fix witjhout tinkering in the code. However, we do see that we have a `ldap.NO_SUCH_OBJECT`.Same thing wether I comment `AUTH_LDAP_APPEND_DOMAIN` or not.

**Comment by dpgaspar** [comment]
Can you check #770

**Comment by AlexisLessard** [comment]
Need to test it ad hoc, but I'm pretty sure that would fix it.

**Comment by thesuperzapper** [comment]
The fix for this issue is found in: #1374
